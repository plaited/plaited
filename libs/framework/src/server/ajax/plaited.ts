import { isTypeOf, canUseDOM } from '@plaited/utils'
import { displayContent } from './display-content.js'
import { DelegatedListener, delegates } from '../../shared/delegated-listener.js'
import { fetchHTML } from './fetch-html.js'
import { NAVIGATE_EVENT_TYPE } from '../../shared/constants.js'

if (canUseDOM()) {
  const navigate = async (event: CustomEvent<URL>) => {
    const { detail: url } = event
    const htmlString = await fetchHTML(url.href)
    if (htmlString) {
      history.pushState(htmlString, '', url.href)
      displayContent(htmlString)
    }
  }
  const pop = ({ state }: PopStateEvent) => {
    if (isTypeOf<string>(state, 'string')) {
      displayContent(state)
    }
  }
  const html = document.querySelector('html') as HTMLHtmlElement
  history.replaceState(new XMLSerializer().serializeToString(html), '', document.location.href)
  !delegates.has(window) && delegates.set(window, new DelegatedListener(pop))
  window.addEventListener('popstate', delegates.get(window))
  !delegates.has(html) && delegates.set(html, new DelegatedListener(navigate))
  html.addEventListener(NAVIGATE_EVENT_TYPE, delegates.get(html))
}

