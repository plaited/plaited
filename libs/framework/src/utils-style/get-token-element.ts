import { defineTemplate } from '../client/define-template.js'
import { css } from '../client/css.js'
import { h } from '../jsx/create-template.js'
import type { CustomElementTag } from '../jsx/types.js'

export const getTokenElement = (stylesheet: string, tag: CustomElementTag = 'design-tokens') => {
  return defineTemplate({
    tag,
    shadowDom: h('slot', {
      stylesheet: css.assign(
        css.host({
          display: 'contents',
        }),
        { stylesheet },
      ).stylesheet,
    }),
  })
}
