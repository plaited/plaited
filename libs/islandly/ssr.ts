import { Template } from './create-template.ts'
export const ssr = (...templates: Template[]) => {
  return '<!DOCTYPE html> ' + templates.map((tpl) => tpl.content).join('')
}
