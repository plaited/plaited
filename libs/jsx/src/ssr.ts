import { TemplateObject } from '@plaited/component-types'
import { escape, isTypeOf } from '../../../libs-cross-framework/utils/dist/index.js'
import { validPrimitiveChildren } from './constants.js'
if (typeof global.HTMLElement === 'undefined') {
  // @ts-ignore node env
  global.HTMLElement = class HTMLElement {}
}

export const ssr = (...templates: TemplateObject[]) => {
  const arr = []
  const stylesheets = new Set<string>()
  const length = templates.length
  for (let i = 0; i < length; i++) {
    const child = templates[i]
    if (isTypeOf<Record<string, unknown>>(child, 'object') && child.$ === 'ðŸ¦„') {
      arr.push(...child.server)
      for (const sheet of child.stylesheets) {
        stylesheets.add(sheet)
      }
      continue
    }
    if (!validPrimitiveChildren.has(typeof child)) continue
    const safeChild = escape(`${child}`)
    arr.push(safeChild)
  }
  const style = stylesheets.size ? `<style>${[...stylesheets].join('')}</style>` : ''
  const str = arr.join('')
  const headIndex = str.indexOf('</head>')
  const bodyRegex = /<body\b[^>]*>/i
  const bodyMatch = bodyRegex.exec(str)
  const bodyIndex = bodyMatch ? bodyMatch.index + bodyMatch[0].length : 0

  const index = headIndex !== -1 ? headIndex : bodyIndex

  return str.slice(0, index) + style + str.slice(index)
}
