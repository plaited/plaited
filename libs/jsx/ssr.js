export * from './dist/ssr.js'
