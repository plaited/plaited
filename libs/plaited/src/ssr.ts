export { ssr } from './jsx/ssr.js'
