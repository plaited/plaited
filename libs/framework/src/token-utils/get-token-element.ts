import { Component, css, h, assignStyles } from 'plaited'

export const getTokenElement = (stylesheet: string, tag: `${string}-${string}` = 'design-tokens') => {
  return Component({
    tag,
    template: h('slot', {
      stylesheet: assignStyles(
        css`
          :host {
            display: contents;
          }
        `,
        { stylesheet },
      ).stylesheet,
    }),
  })
}
