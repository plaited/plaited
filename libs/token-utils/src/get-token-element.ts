import { Component, css, h } from 'plaited'
import { stylesheets } from 'plaited/css'

const { $stylesheet } = css`
  :host {
    display: contents;
  }
`

export const getTokenElement = (stylesheet: string, tag: `${string}-${string}` = 'design-tokens') => {
  return Component({
    tag,
    template: h('slot', { stylesheet: stylesheets($stylesheet, stylesheet) }),
  })
}
