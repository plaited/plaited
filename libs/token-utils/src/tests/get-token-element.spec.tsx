import { test, expect } from 'bun:test'
import { getTokenElement } from '../get-token-element.js'
import { ssr } from 'plaited/ssr'
import beautify from 'beautify'

test('getTokenElement', () => {
  let DesignTokens = getTokenElement(`:host{--blue: blue}`)

  expect(beautify(ssr(<DesignTokens />), { format: 'html' })).toMatchSnapshot()
  DesignTokens = getTokenElement(`:host{--blue: blue}`, 'plaited-tokens')
  expect(
    beautify(
      ssr(
        <DesignTokens>
          <div>hello world!</div>
        </DesignTokens>,
      ),
      { format: 'html' },
    ),
  ).toMatchSnapshot()
})
