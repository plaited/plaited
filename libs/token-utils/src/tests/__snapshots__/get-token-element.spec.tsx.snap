// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenElement 1`] = `
"<design-tokens>
    <template shadowrootmode="open" shadowrootdelegatesfocus="true">
        <style>
            :host { display: contents; }:host { --blue: blue; }
        </style>
        <slot></slot>
    </template>
</design-tokens>"
`;

exports[`getTokenElement 2`] = `
"<plaited-tokens>
    <template shadowrootmode="open" shadowrootdelegatesfocus="true">
        <style>
            :host { display: contents; }:host { --blue: blue; }
        </style>
        <slot></slot>
    </template>
    <div>hello world!</div>
</plaited-tokens>"
`;
